### PR TITLE
Add option to fall back on regex parsing to splitAddress1

### PR DESCRIPTION
**Pull request by gabypancu**
### What are you trying to accomplish?

Resolves https://github.com/Shopify/address/issues/2726

When there is no reserved delimiter present in an address line (i.e. any address that was created before the release of additional address fields), we want to attempt to split the additional fields out using regex. The regex patterns will have to be strict in order to minimize incorrect captures.

### What approach did you choose and why?

1. Adds a conservative address1 regex pattern for the Netherlands. On a sample of about 800 address lines, it parses 75% correctly (the rest are not matched) - https://rubular.com/r/3BNDUviIL8hHeS
~~2. Adds a `splitAddress1WithFallback` function to the TS library that uses the regex as a fallback if splitting on the delimiter is not possible~~
2. Adds an optional param `tryRegexFallback` to the existing `splitAddress1` function in the TS library. When true, and when the input string does not contain the reserved delimiter, we attempt to parse using regex.

This way, checkout web still only has one function to call for splitting.

The following PRs have been merged into this one:
- https://github.com/Shopify/worldwide/pull/282
- https://github.com/Shopify/worldwide/pull/286

**See PR comments for debate on the following pros and cons**
Our options for where to implement this new behaviour:
1. In the existing `splitAddress1` function, controlled by a new optional param
- Requires the smallest change in checkout-web, but brings higher visibility to this feature
- Reusability if we choose to use this on surfaces other than checkout-web
2. In a new function in Worldwide
- Maintains separation with the existing splitting function, reduces visibility to this feature
- Reusability if we choose to use this on surfaces other than checkout-web
3. Directly in checkout-web when splitting fails
- Keeps this feature private to Shopify
- Introduces complexity & Worldwide domain into checkout-web
- Not reusable outside of checkout-web

### Checklist

* [x] I have added a CHANGELOG entry for this change (or determined that it isn't needed)

### CODE REVIEW

File: db/data/regions/NL.yml
Comment (rochlefebvre) [reply db/data/regions/NL.yml:33]: `\s` always matches the space, tab, newline, carriage return, and form feed characters. I believe Javascript also includes unicode flavours as well.

How about we be more restrictive and just match on literal space characters?

```regex
^(?<streetName>[^\\d]+) (?<streetNumber>\\d+(?: ?[A-Za-z])?)$
```

File: lang/typescript/src/extended-address/splitAddress1WithFallback.test.ts
Comment (kennygoff) [inline lang/typescript/src/extended-address/splitAddress1WithFallback.test.ts:80]: For this and the above test that handles multiple test cases Vitest has a utility called [`test.each`](https://vitest.dev/api/#test-each). You can pass the inputs as an object to `test.each` and provide a clean way to print the results for each test case. I'm not sure if you can pass objects as input though so you might have to do flatten the input objects:

```ts
// Nested objects might work, but I'm not 100% sure
test.each([
  { input: { country: 'NL', address: 'Kempenaar 25 11' }, expected: 'Kempenaar 25 11' }
])('splitAddress1WithFallback($input.country, $input.country) -> $expected', ({ input, expected }) => {
  expect(splitAddress1WithFallback(input.country, input.address)).toBe(expected)
})

// Flattened, definitely will work
test.each([
  { country: 'NL', address: 'Kempenaar 25 11', expected: 'Kempenaar 25 11' }
])('splitAddress1WithFallback($country, $address) -> $expected', ({ country, address, expected }) => {
  expect(splitAddress1WithFallback(country, address)).toBe(expected)
})
```

File: lang/typescript/src/utils/address-fields.ts
Comment (kennygoff) [inline lang/typescript/src/utils/address-fields.ts:39]: This function wasn't updated so this doc description shouldn't change

File: lang/typescript/src/extended-address/splitAddress1WithFallback.ts
Comment (kennygoff) [inline lang/typescript/src/extended-address/splitAddress1WithFallback.ts:20]: @gabypancu I actually think I prefer just changing the default behavior or making it an optional param to turn on/off the regex fallback. Not a super strong opinion so if you have a good reason this is fine, but the library is new enough and only has one main consumer so I feel like making the decision now to change it to add this fallback always if the user wants to opt-out. At the end of the day all of the split logic is defined in the region yamls so I think we can safely change this and just bump the major version to signify the clear change in scope.

I don't hate just doing `splitAddress1('NL', 'Main street 123', false)` even if its a little clunky if we want to offer a way to turn off the fallback regex.

If we go with a second function we should document this in the README too.

File: db/data/regions/NL.yml
Comment (gabypancu) [reply db/data/regions/NL.yml]: Good callout 👍 done!

File: lang/typescript/src/utils/address-fields.ts
Comment (kennygoff) [inline lang/typescript/src/utils/address-fields.ts:80]: Missing the first param for fieldDefinition here `@param fieldDefinition Array of definitions of address sub-fields`

**Review by kennygoff — CHANGES_REQUESTED** [review]

File: lang/typescript/src/extended-address/splitAddress1WithFallback.ts
Comment (gabypancu) [reply lang/typescript/src/extended-address/splitAddress1WithFallback.ts:20]: This was initially my preference as well. @rochlefebvre brought up a good point that this "fallback" logic is not ideal, and we don't want to draw attention to it. That was the main motivator for having it in a separate function.

File: lang/typescript/src/utils/address-fields.ts
Comment (gabypancu) [reply lang/typescript/src/utils/address-fields.ts:80]: Thanks for catching that!

File: lang/typescript/src/extended-address/splitAddress1WithFallback.ts
Comment (kennygoff) [reply lang/typescript/src/extended-address/splitAddress1WithFallback.ts:20]: Yeah just run the pnpm changeset is all. I was recommending to update the README to document usage of the new function but if we want it to be undocumented on purpose then no need. After it gets merged to main a PR will open to handle the release.

Comment (kennygoff) [reply lang/typescript/src/extended-address/splitAddress1WithFallback.ts:20]: I think my main concern right now is that this is still a public facing function but it's not really clear why it's there when reading through it without context. So if someone goes to import the documented `splitAddress1` function they will see a semi-documented second `splitAddress1WithFallback` function with a similar name in autocomplete/IntelliSense. I think this might have the opposite effect we want: "Oh I should use the one with the fallback, because it has fallback functionality which is a good thing". We could double-down on the "don't use this" or "it's experimental" and just popping a warning in the docstring comment for the new fallback function.

Imho though I think adding it as an explicit set of options on the existing function by doing `splitAddress1('NL', 'Main 123', { experimentalRegexFallback: true })`. That way it's opt-in and very clear it's potentially not something someone should use without knowing what they're doing. Then if don't want to document it in the README or later we decide we want to no longer expose this option on the public surface of the API we can just remove it.

Comment (gabypancu) [reply lang/typescript/src/extended-address/splitAddress1WithFallback.ts]: Discussed off-Github - we've aligned on moving the fallback logic into the existing `splitAddress1` function behind an optional `experimentalRegexFallback` flag for the reasons outlined above. We will also release this as a beta feature, given that support for countries will be added incrementally

File: lang/typescript/src/extended-address/splitAddress1.ts
Comment (gabypancu) [inline lang/typescript/src/extended-address/splitAddress1.ts:28]: Open to other names. I considered `tryRegexFallback` as it's more concise and suggests that the splitting may not be successful

File: lang/typescript/src/extended-address/splitAddress1WithFallback.test.ts
Comment (gabypancu) [reply lang/typescript/src/extended-address/splitAddress1WithFallback.test.ts]: Oh nice! Applied 👍

File: lang/typescript/src/extended-address/splitAddress1.ts
Comment (kennygoff) [reply lang/typescript/src/extended-address/splitAddress1.ts:28]: `try` as a prefix works, although I think `experimental` feels more "dangerous" which seems like it's what we want to convey. Up to you, I'm not too concerned with the specific prefix though as long as it seems to make sense when reading. Just don't use `useRegexFallback` as that prefix has a specific use case in react and could confuse consumers.

File: lang/typescript/src/extended-address/splitAddress1.ts
Comment (kennygoff) [inline lang/typescript/src/extended-address/splitAddress1.ts:37]: This will change existing behavior for parsing on countries that do not have a config thats untested right now. I'll suggest a code edit to include a test case to catch this behavior.

```ts
splitAddress1('US', '')) // returns null today
splitAddress1('US', '')) // returns {} with this change
```

We might want to invert this big if statement below and just return early:

```ts
if(!fieldConcatenationRules) {
  return null;
}

if (address === '') {
  return {};
}

// Now do the rest of the important config checks not in an if block
```

File: lang/typescript/src/extended-address/splitAddress1.test.ts
Comment (kennygoff) [inline lang/typescript/src/extended-address/splitAddress1.test.ts:6]: The current behavior is to return null if the country has no region config, regardless of what is inputted. Here's some new test cases that'll cover that.

```suggestion
    expect(splitAddress1('US', '123 Main')).toBeNull();
    expect(splitAddress1('US', '')).toBeNull();
    expect(splitAddress1('US', '123 Main', true)).toBeNull();
    expect(splitAddress1('US', '', true)).toBeNull();
```

File: lang/typescript/src/utils/address-fields.ts
Comment (kennygoff) [inline lang/typescript/src/utils/address-fields.ts:83]: Can you also add a test for this function in `address-fields.test.ts`?

File: lang/typescript/CHANGELOG.md
Comment (kennygoff) [reply lang/typescript/CHANGELOG.md:8]: Looks like it was but I don't see the package published as a prerelease on NPM so might have to change something in the GH Actions to actually do it https://www.npmjs.com/package/@shopify/worldwide?activeTab=versions

Comment (kennygoff) [reply lang/typescript/CHANGELOG.md:8]: Ok yeah changesets does recommend using a second branch when doing prereleases just so the main branch isn't locked up in prerelease mode. Given we don't make that many changes to this library and we won't be in prerelease for long thats probably still fine to just lock up the main branch but I'd rather not if we don't have to so `main` stays in a releasable state that reflects the latest version.

We can try to run the publish on this branch, we'll need to modify the [GH action](https://github.com/Shopify/worldwide/blob/main/.github/workflows/npm-release.yml) to also run on this branch though so the job that handles publishing will run.

File: lang/typescript/src/extended-address/splitAddress1.ts
Comment (gabypancu) [reply lang/typescript/src/extended-address/splitAddress1.ts:37]: Ah thanks for catching that edge case!

File: lang/typescript/src/utils/address-fields.ts
Comment (gabypancu) [reply lang/typescript/src/utils/address-fields.ts:83]: Done!

**Review by kennygoff — APPROVED** [review]
LGTM 👍 If we want to do the prerelease I still recommend trying to publish the prerelease from this branch instead of merging though as the changesets docs recommend.

File: lang/typescript/src/utils/address-fields.test.ts
Comment (DominiqueFlaaa) [inline lang/typescript/src/utils/address-fields.test.ts:227]: nit: can you update this test to have 2 regexPatterns, with the first not a match and the second a match?

File: .github/workflows/npm-release.yml
Comment (gabypancu) [inline .github/workflows/npm-release.yml:7]: TODO: remove once done with pre-release

File: lang/typescript/CHANGELOG.md
Comment (DominiqueFlaaa) [reply lang/typescript/CHANGELOG.md:27]: >The pre-releases are published and available in [npm](https://www.npmjs.com/package/@shopify/worldwide/v/0.7.0-next.0)

Ah thank you for clarifying.

**Pull request merged**
